packages/eslint-plugin/src/rules/strict-boolean-expressions.ts

```diff
@@ -260,19 +268,124 @@ export default createRule<Options, MessageId>({
       // if the logical expression is used for control flow,
       // then its right argument is used for its side effects only
-      traverseNode(node.right, isCondition);
+      const isRightReported = traverseNode(node.right, isCondition);
+
+      return isLeftReported || isRightReported;
+    }
+
+    /**
+     * Inspects return statements of predicate functions.
+     */
+    function traverseArrayPredicateReturnStatements(
+      node: TSESTree.CallExpressionArgument,
+    ): { hasReported: boolean; hasReturnStatement: boolean } {
+      let hasReturnStatement = false;
+      let hasReported = false;
+
+      // Shorthand arrow function expression
+      if (
+        node.type === AST_NODE_TYPES.ArrowFunctionExpression &&
+        node.body.type !== AST_NODE_TYPES.BlockStatement
+      ) {
+        hasReturnStatement = true;
+        hasReported = traverseNode(node.body, true);
+      }
+
+      // Any other function expression with a block body
+      else if (
+        (node.type === AST_NODE_TYPES.FunctionExpression ||
+          node.type === AST_NODE_TYPES.ArrowFunctionExpression) &&
+        node.body.type === AST_NODE_TYPES.BlockStatement
+      ) {
+        const tsBody = services.esTreeNodeToTSNodeMap.get(node.body);
+
+        forEachReturnStatement(tsBody, tsStatement => {
+          hasReturnStatement = true;
+
+          const statement = services.tsNodeToESTreeNodeMap.get(tsStatement);
+
+          if (statement.type === AST_NODE_TYPES.ReturnStatement) {
+            if (statement.argument) {
+              hasReported ||= traverseNode(statement.argument, true);
+            } else {
+              hasReported = true;
+
+              context.report({
+                node: statement,
+                messageId: 'predicateReturnsEmptyExpression',
+              });
+            }
+          }
+        });
+      }
+
+      return { hasReported, hasReturnStatement };
     }
 
     function traverseCallExpression(node: TSESTree.CallExpression): void {
       const assertedArgument = findTruthinessAssertedArgument(services, node);
       if (assertedArgument != null) {
         traverseNode(assertedArgument, true);
       }
+      if (isArrayMethodCallWithPredicate(context, services, node)) {
+        const predicate = node.arguments.at(0);
+
+        if (predicate != null) {
+          const { hasReported, hasReturnStatement } =
+            traverseArrayPredicateReturnStatements(predicate);
+
+          // Don't report a function as implicitly returning `undefined` only
+          // when no other issues were found with it.
+          if (hasReported) {
+            return;
+          }
+
+          // Empty functions has their return type as `void` rather than `undefined`
+          if (
+            !hasReturnStatement &&
+            (predicate.type === AST_NODE_TYPES.FunctionExpression ||
+              predicate.type === AST_NODE_TYPES.ArrowFunctionExpression)
+          ) {
+            return context.report({
```

Comment:
In cases where we know it's a function literal, what do you think of adding a suggestion to put an explicit `: boolean` return type? Then the user can let the type errors guide them in resolving which return statements are illegal, rather than us having to do a best effort to underline them?

Reply:
Nice! I'll update the PR with this shortly.

Reply:
@kirkwaiblinger this may collide with #9764 (at least for array's`.filter()`), in which having an explicit boolean return type may prevent TypeScript from inferring the predicate return type.

Wdyt? I'm not sure if this is correct (at least for `.filter()`).

Reply:
Hmmmmmm. Well, it's not a hard collision, in that _if_ an explicit type predicate is inferred, then we wouldn't be reporting the function in the first place, since, to be an inferred type predicate, the function must be returning a boolean.

I see what you're saying that it's a bit of an awkward workflow, in that one rule might ask you to put `: boolean`, just for another to ask you to take it away.
```ts
declare const arr: Array<string | null>;

arr.filter(x => x); // flagged by SBE

// suggested to
arr.filter((x): boolean => x); // TS ERROR, x not assignable to boolean, _and_ lint error, remove explicit `: boolean`

// manually fix to
arr.filter((x): boolean => x != null); // lint error, remove explicit `: boolean`

// OR manually fix to
arr.filter((x): boolean => x != null && x !== ''); // lint error, remove explicit `: boolean`
// _but_ note that this would be a false positive since a type predicate is _not_ inferred here.
```

But... since it's a _suggestion_ TBH I think it's ok. Suggestions can always suggest things that another rule may disapprove of, and they should not generally be understood to be the only way to resolve a report. The goal is to guide you to finding a solution, which, IMO, this would do (by surfacing TS type errors), even if it's a little awkward. In any case that rule doesn't exist yet, so 🤷 I don't think we should stress about it too much.

***

That does remind me, though, if there isn't already, could you ensure there's a (valid) test case for explicit and inferred type predicates, to be sure we don't flag them as non-boolean (it did work as expected when I tried the playground 👍)

Reply:
definitely a good observation, though! Really appreciate the deep attention to detail.